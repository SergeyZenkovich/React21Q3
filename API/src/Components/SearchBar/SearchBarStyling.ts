import styled from "styled-components";
import fontP from "../../assets/fonts/MavenPro-Regular.ttf";

export const StyledSerachBlockWithOptions = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 140px;
  background-color: #000a;
  margin-bottom: 10px;
  padding: 0;
`;
export const StyledSearchBlock = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 70px;
  padding: 0;
`;
export const StyledoptionsBlock = styled.div`
  @font-face {
    font-family: "CardTextFont";
    src: url(${fontP});
  }
  display: flex;
  justify-content: center;

  & select {
    margin: 10px 0;
    height: 40px;
    font-family: "CardTextFont";
    font-size: 20px;
    border: none;
    outline: none;
  }
`;
export const StyledSearchInput = styled.input`
  @font-face {
    font-family: "CardTextFont";
    src: url(${fontP});
  }
  display: flex;
  width: 75%;
  height: 45px;
  padding: 0 10px;
  font-family: "CardTextFont";
  outline: none;
  border: none;
  border-radius: 3px;
  font-size: 1em;
`;
export const StyledCardButton = styled.button`
  @font-face {
    font-family: "CardTextFont";
    src: url(${fontP});
  }
  width: 15%;
  height: 45px;
  margin: 0 5px;
  font-family: "CardTextFont";
  border: none;
  font-size: 1em;
  border-radius: 3px;
  background-color: #ddde;
  &:hover {
    cursor: pointer;
    background-color: #eeef;
  }
`;
