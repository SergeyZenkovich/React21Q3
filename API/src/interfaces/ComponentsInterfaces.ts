export interface CardInterface {
  cardProblem: string;
  cardDate: string;
  cardPriority: string;
  cardisRequestNeeded: boolean;
}
export interface ImageCardInterface {
  url: string;
}
export interface FetchObject {
  alt_description: string;
  color: string;
  id: string;
  likes: number;
  urls: FetchedURLS;
  user: FetchedUser;
}
export interface URLs {
  urls?: {
    full: string;
    regular: string;
    small: string;
    thumb: string;
  };
}
export interface FetchedURLS {
  full: string;
  regular: string;
  small: string;
  thumb: string;
}
export interface FetchedUser {
  first_name: string;
  last_name: string;
  location: string;
  username: string;
  twitter_username: string;
  instagram_username: string;
  total_photos: number;
  total_likes: number;
}
export interface NavElement {
  page: string;
  link: string;
}
export interface QueryParametersInterface {
  query: string;
  page: number;
  elementsOnPage: number;
  orient: string;
  color: string;
  orderBy: string;
}
export interface IdParameterInterface {
  id: string;
}
export interface RequestResponseInterface {
  results: FetchObject[];
  total_pages: number;
}
