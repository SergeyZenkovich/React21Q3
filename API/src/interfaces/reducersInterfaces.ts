import {
  FetchedURLS,
  FetchedUser,
  FetchObject,
  QueryParametersInterface
} from "./ComponentsInterfaces";

/// home page reducer

export interface HomePageStateInterface {
  totalElementsCount: number;
  elements: FetchObject[];
  isFetching: boolean;
  requestParams: QueryParametersInterface;
}

export enum FetchActionsType {
  TOGGLE_FETCHING = "TOGGLE-LOADING",
  SET_TOTAL_COUNT = "SET-TOTAL-COUNT",
  SET_ELEMENTS = "SET-ELEMENTS",
  SET_PAGE_PARAM = "SET-PAGE-PARAM",
  SET_QUERY_PARAM = "SET-QUERY-PARAM",
  SET_ORIENT_PARAM = "SET-ORIENT-PARAM",
  SET_COLOR_PARAM = "SET-COLOR-PARAM",
  SET_OREDER_BY_PARAM = "SET-ORDER-BY-PARAM",
  SET_ELEMENTS_ON_PAGE_PARAM = "SET-ELEMENTS-ON-PAGE-PARAM",
  SET_PARAMS = "SET-PARAMS"
}

export interface LoadingAction {
  type: FetchActionsType.TOGGLE_FETCHING;
}
export interface SetTotalCountAction {
  type: FetchActionsType.SET_TOTAL_COUNT;
  totalCount: number;
}
export interface SetElementsAction {
  type: FetchActionsType.SET_ELEMENTS;
  elements: FetchObject[];
}
export interface SetPageParamAction {
  type: FetchActionsType.SET_PAGE_PARAM;
  page: number;
}
export interface SetQueryParamAction {
  type: FetchActionsType.SET_QUERY_PARAM;
  query: string;
}
export interface SetOrientParamAction {
  type: FetchActionsType.SET_ORIENT_PARAM;
  orient: string;
}
export interface SetColorParamAction {
  type: FetchActionsType.SET_COLOR_PARAM;
  color: string;
}
export interface SetOrderByParamAction {
  type: FetchActionsType.SET_OREDER_BY_PARAM;
  orderBy: string;
}
export interface SetElementsOnPageParamAction {
  type: FetchActionsType.SET_ELEMENTS_ON_PAGE_PARAM;
  elementsOnPage: number;
}
export interface SetParamsAction {
  type: FetchActionsType.SET_PARAMS;
  params: QueryParametersInterface;
}

export type FetchActions =
  | LoadingAction
  | SetTotalCountAction
  | SetElementsAction
  | SetParamsAction
  | SetPageParamAction
  | SetQueryParamAction
  | SetOrientParamAction
  | SetColorParamAction
  | SetOrderByParamAction
  | SetElementsOnPageParamAction;

/// details page reducer

export interface DetailsPageStateInterface {
  isLoading: boolean;
  element: {
    id: string;
    color: string;
    alt_description: string;
    likes: number;
    urls: FetchedURLS;
    user: FetchedUser;
  };
}

export enum FetchActionsTypeDetails {
  TOGGLE_FETCHING = "TOGGLE-LOADING",
  SET_ELEMENT = "SET-ELEMENT"
}

export interface SetDetailsElementAction {
  type: FetchActionsTypeDetails.SET_ELEMENT;
  element: FetchObject;
}
export interface LoadingDetailsAction {
  type: FetchActionsTypeDetails.TOGGLE_FETCHING;
}

export type DetailsFetchingActions = SetDetailsElementAction | LoadingDetailsAction;
