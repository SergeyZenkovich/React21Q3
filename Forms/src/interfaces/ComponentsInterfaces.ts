export interface CardInterface {
  cardProblem: string;
  cardDate: string;
  cardPriority: string;
  cardisRequestNeeded: boolean;
  production: boolean;
}
