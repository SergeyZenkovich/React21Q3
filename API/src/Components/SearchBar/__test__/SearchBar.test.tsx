import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import styled from "styled-components";
import "jest-styled-components";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import Search from "../SearchBar";
import {StyledSearchInput} from "../SearchBarStyling";
import store from "../../../redux/store";
import {
  setRequestOrderParam,
  setRequestQueryParam
} from "../../../redux/ActionCreators/homePageActionCreators";
import {FetchActionsType} from "../../../interfaces/reducersInterfaces";

const middlewares = [];
const mockStore = configureStore(middlewares);

const testObj = {
  query: "testQuery",
  searchData: jest.fn(),
  orderBy: "relevant",
  orient: "landscape",
  color: "blue"
};

it("renders img correctly", () => {
  const container = document.createElement("div");
  ReactDOM.render(
    <Provider store={store}>
      <Search {...testObj}></Search>
    </Provider>,
    container
  );

  expect(container.querySelector("input")).toBeTruthy();
});
it("renders input with query correctly", () => {
  const container = document.createElement("div");
  ReactDOM.render(
    <Provider store={store}>
      <Search {...testObj}></Search>
    </Provider>,
    container
  );
  const innerBlockClassName = StyledSearchInput.styledComponentId;
  const innerBlock = container.querySelector(`.${innerBlockClassName}`) as HTMLInputElement;
  expect(innerBlock.value).toEqual(testObj.query);
});
it("renders select with value correctly", () => {
  const container = document.createElement("div");
  ReactDOM.render(
    <Provider store={store}>
      <Search {...testObj}></Search>
    </Provider>,
    container
  );
  const OrderSelectElement = container.querySelector(`select[id="orderBy"]`) as HTMLSelectElement;
  const OrientSelectElement = container.querySelector(
    `select[id="orientation"]`
  ) as HTMLSelectElement;
  const ColorSelectElement = container.querySelector(`select[id="color"]`) as HTMLSelectElement;
  expect(OrderSelectElement.value).toEqual(testObj.orderBy);
  expect(OrientSelectElement.value).toEqual(testObj.orient);
  expect(ColorSelectElement.value).toEqual(testObj.color);
});

describe("SearchBar Component", () => {
  const setFunction = jest.fn();
  let container;
  beforeEach(() => {
    container = renderer.create(
      <Provider store={store}>
        <Search {...testObj}></Search>
      </Provider>
    );
  });
  it("dispacth correctly", () => {
    const initialState = {};
    const storeM = mockStore(initialState);
    storeM.dispatch(setRequestQueryParam("test"));
    const actions = storeM.getActions();
    const expectedPayload = {
      type: FetchActionsType.SET_QUERY_PARAM,
      query: "test"
    };
    expect(actions).toEqual([expectedPayload]);
  });
  it("dispacth correctly", () => {
    const initialState = {};
    const storeM = mockStore(initialState);
    storeM.dispatch(setRequestOrderParam("latest"));
    const actions = storeM.getActions();
    const expectedPayload = {
      type: FetchActionsType.SET_OREDER_BY_PARAM,
      orderBy: "latest"
    };
    expect(actions).toEqual([expectedPayload]);
  });
  it("matches shapshot", () => {
    const tree = container.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
