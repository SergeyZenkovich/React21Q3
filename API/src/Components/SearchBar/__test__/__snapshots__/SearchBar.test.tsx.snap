// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Component matches shapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  height: 140px;
  background-color: #000a;
  margin-bottom: 10px;
  padding: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 70px;
  padding: 0;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

.c4 select {
  margin: 10px 0;
  padding: 0 10px;
  height: 40px;
  font-family: "CardTextFont";
  font-size: 20px;
  border: none;
  outline: none;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 75%;
  height: 45px;
  padding: 0 10px;
  font-family: "CardTextFont";
  outline: none;
  border: none;
  border-radius: 3px;
  font-size: 1em;
}

.c3 {
  width: 15%;
  height: 45px;
  margin: 0 5px;
  font-family: "CardTextFont";
  border: none;
  font-size: 1em;
  border-radius: 3px;
  background-color: #ddde;
}

.c3:hover {
  cursor: pointer;
  background-color: #eeef;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <input
      className="c2"
      onChange={[Function]}
      placeholder="Iphone X...."
      value="testQuery"
    />
    <button
      className="c3"
      onClick={[Function]}
    >
      Search
    </button>
  </div>
  <div
    className="c4"
  >
    <select
      id="orderBy"
      name="orderBy"
      onChange={[Function]}
      value="relevant"
    >
      <option
        value="relevant"
      >
        Relevant
      </option>
      <option
        value="latest"
      >
        Latest
      </option>
    </select>
    <select
      id="orientation"
      name="orientation"
      onChange={[Function]}
      value="landscape"
    >
      <option
        value="landscape"
      >
        Landscape
      </option>
      <option
        value="portrait"
      >
        Portrait
      </option>
      <option
        value="squarish"
      >
        Squarish
      </option>
    </select>
    <select
      id="color"
      name="color"
      onChange={[Function]}
      value="blue"
    >
      <option
        value="black_and_white"
      >
        Black and White
      </option>
      <option
        value="black"
      >
        Black
      </option>
      <option
        value="white"
      >
        White
      </option>
      <option
        value="yellow"
      >
        Yellow
      </option>
      <option
        value="orange"
      >
        Orange
      </option>
      <option
        value="red"
      >
        Red
      </option>
      <option
        value="purple"
      >
        Purple
      </option>
      <option
        value="magenta"
      >
        Magenta
      </option>
      <option
        value="green"
      >
        Green
      </option>
      <option
        value="teal"
      >
        Teal
      </option>
      <option
        value="blue"
      >
        Blue
      </option>
    </select>
  </div>
</div>
`;
